### ISSUE TITLE
Save m4b with chapters destroys the headers and makes the file unreadable.

### ISSUE THREAD

**Issue by ElmoFuntz** [Zeugma440/atldotnet#47]
## The problem

Wrote test data to a smaller sized m4b and it appears to have wiped any meta data and made the file unreadable.

I have sent you a copy of the file i am testing on. 

Ok I also just checked the larger file I was testing last night and while it did not crash the app and I am able to read the metadata back the file is also corrupt. This one plays but just a quick humming tone then ends.

Edit -- 

I did some additional testing and stripped the chapters out of the file i sent you and was able to update/save the file without it crashing but the file is still unplayable (I actually got one test to play but no sound). This appears to be related to having (or had) chapter data since an m4a that has not had chapter data works (except for bug #49).

## Environment

Using version 2.16

## Details

Logging info:

Upon reading the file in:

Title : Test file
Artist : 
Album : Test author
Recording year : 2019
Track number : 0
Disc number : 0
Genre : Audiobook
Comment : this is a comment
Duration (s) : 28504
Bitrate (KBps) : 68
Number of channels : 2
Channels arrangement : Stereo (2/0.0)
Has variable bitrate audio : yes
Has lossless audio : no
Custom field ©too : value = Lavf58.17.101

After clicking save and trying to read load the track info:
d:/test.m4b | moov atom could not be found; aborting read [?=1472 ms]
d:/test.m4b | moov atom could not be found; aborting read [?=69615 ms]
d:/test.m4b | moov atom could not be found; aborting read [?=2 ms]
Title : test
Artist : 
Album : 
Recording year : 0
Track number : 0
Disc number : 0
Genre : 
Comment : 
Duration (s) : 0
Bitrate (KBps) : 0
Exception thrown: 'System.NullReferenceException' in M4BCreator.exe
theTrack.**<Field at offset 0x0000005C>** was null.

The line it is erroring on is:

 System.Console.WriteLine("Number of channels : " + theTrack.ChannelsArrangement.NbChannels);

The file is no longer playable.

I am using your example code

        private void btnSave_Click(object sender, EventArgs e)
        {
            // Adapt this to whatever your file path needs to be
            string fileName = "d:/test.m4b";

            // Load audio file information into memory
            Track theTrack = new Track(fileName);

            // Modify metadata
            theTrack.Artist = "Hey ho";
            theTrack.Composer = "Oscar Wilde";

            // Save modifications on the disc
            theTrack.Save();
        }

**Comment by Zeugma440** [comment]
Nice one, thanks for the report. There's an issue with padding management I had not detected until now.

**Comment by ElmoFuntz** [comment]
Very nice. Thanks for the quick fix. This is the one that affects my project the most.

**Comment by Zeugma440** [comment]
Made significant progress with the sample you gave me earlier, but it turns out it contains _both_ Quicktime chapters and Nero chapters. 

One of these chapters table is totally fixed (chapters appear correctly in Media Player and iTunes) and the other is not (chapters appear with wrong time offsets in foobar2000 and VLC).

I'll keep you updated

**Comment by ElmoFuntz** [comment]
Hmm OK. Those tags/chapters read in mp3tag, iTunes, iOS, and windows.  I've got no idea how the internals of audio metadata markup works though. I'm not even sure how to see it has both types but would be cool to know how. Looking at the file in mp3tag it's not even showing the tag is in nero format. Mp3tag does seem to prefer m4b files are saved as nero though since it converts any tags for those to "MP4 Nero." I have not edited and saved that file in mp3tag though so it currently shows an "MP4" format tag when viewing in mp3tag.

**Comment by Zeugma440** [comment]
Since these formats (Quicktime and Nero) lack a good documentation, there is no free software that can easily handle both. Maybe professional editing softwares can ?

If you're curious about it, I've written down the technical details in the wiki : https://github.com/Zeugma440/atldotnet/wiki/Focus-on-Chapter-metadata

Right now, and regardless of the bug we're discussing here, ATL can read both chapter formats but only supports writing Nero chapters. Writing QT chapters is super complex as their data is interleaved with audio chunks.

**Comment by ElmoFuntz** [comment]
Cool thanks for the info. I'm wondering how QT chapters even got written in that file. OTOH i'm not positive that file is one I converted and wrote the original chapters for. Thanks for all your work on this!

**Comment by ElmoFuntz** [comment]
I did some more testing by encoding a new file from a set of mp3 files and after a bit of digging around it appears ffmpeg is in fact writing both of these chapter formats which I was able to see in mediainfo (if i'm reading this correctly). There was also a post with no answer on the stack forums asking how to prevent this.  https://video.stackexchange.com/questions/28068/ffmpeg-nero-and-quicktime-format-chapters

**Comment by Zeugma440** [comment]
` Is there an "easy" way to tell what formats are in a file?`

Yes there is. Just open an hexadecimal editor such as [FrHed](https://sourceforge.net/projects/frhed/), and looks for the following atom identifiers
  - `chap` if present, there are QT chapters in your file
  - `chpl` if present, there are Nero chapters in your file

***
`after a bit of digging around it appears ffmpeg is in fact writing both of these chapter formats`

oh, this explains that. Thanks for digging this up !

**Comment by Zeugma440** [comment]
Done ! 

That one took a long time because the MP4/M4A class had a lot of unchecked issues that prevented it from properly editing files with multiple tracks and missing nodes.
Now everything should be in order.

=> Here's the test library with fixes for all 4 issues you reported

[ATL.zip](https://github.com/Zeugma440/atldotnet/files/4105943/ATL.zip)

Could you try it and report any remaining issue ? I'll publish it on nuGet once everything is cleared~

**Comment by ElmoFuntz** [comment]
Thanks! I will give it a shot this evening after work.

**Comment by ElmoFuntz** [comment]
Man I had hopes for this one but it is still doing the same thing

Load m4b with chapters (it has both types)
Save file and items are zero'd out again

Log from load and save
```
Duration (s) : 64899
Bitrate (KBps) : 63
Number of channels : 2
Channels arrangement : Stereo (2/0.0)
Has variable bitrate audio : yes
Has lossless audio : no
Custom field ©too : value = Lavf55.45.100
d:/test.m4b | Disk stream operation : Lengthening (delta=71710) [?=32859 ms]
d:/test.m4b | moov atom could not be found; aborting read [?=2536 ms]
```

Try and reload and it crashes reading channel count

```
d:/test.m4b | moov atom could not be found; aborting read [?=259579 ms]
d:/test.m4b | moov atom could not be found; aborting read [?=1 ms]
Title : test
Artist : 
Album : 
Recording year : 0
Track number : 0
Disc number : 0
Genre : 
Comment : 
Duration (s) : 0
Bitrate (KBps) : 0
Exception thrown: 'System.NullReferenceException' in M4BCreator.exe
theTrack.**<Field at offset 0x0000005C>** was null.

```

Comment out the read for channels and it does not crash but all values are zero

```
d:/test.m4b | moov atom could not be found; aborting read [?=2717 ms]
d:/test.m4b | moov atom could not be found; aborting read [?=19 ms]
Title : test
Artist : 
Album : 
Recording year : 0
Track number : 0
Disc number : 0
Genre : 
Comment : 
Duration (s) : 0
Bitrate (KBps) : 0
Has variable bitrate audio : no
Has lossless audio : no
```

Code

        private void displayProgress(float progress)
        {
            txtProgress.Text = progress + "%";
        }

         private void btnSave_Click(object sender, EventArgs e)
        {
            // Adapt this to whatever your file path needs to be
            string fileName = "d:/test.m4b";

            IProgress<float> progress = new Progress<float>(displayProgress);
            // Load audio file information into memory
            Track theTrack = new Track(fileName, progress);

            // Modify metadata
            theTrack.Artist = "Hey ho";
            theTrack.Composer = "Oscar Wilde";
            theTrack.Album = "Fake album starts here and is longer than the original one";

            // Save modifications on the disc

            theTrack.Save();
        }

**Comment by Zeugma440** [comment]
I confirm I can repro it on my side. Sorry for the false hopes. I'll get back on it and keep you informed.

**Comment by ElmoFuntz** [comment]
Thanks. I really appreciate all the work you are doing on this!

**Comment by Zeugma440** [comment]
Found the issue. During the refactoring I made last week, I've associated an atom to the wrong parent in the file structure hierarchy. Now it works !

I'll send a new build once #48 is fixed.

**Comment by ElmoFuntz** [comment]
Awesome!

**Comment by ElmoFuntz** [comment]
Something still wrong.  Save file and get the below. When i reload all metadata is gone again and the file does not play. This file did start with both chapter types (all 500+ of my files appear to have both).

```
Bitrate (KBps) : 63
Has variable bitrate audio : yes
Has lossless audio : no
Custom field ©too : value = Lavf55.45.100
d:/test.m4b | trak atom could not be found for index 3; aborting reading through tracks [?=11022 ms]
d:/test.m4b | trak atom could not be found for index 3; aborting reading through tracks [?=59 ms]
d:/test.m4b | trak atom could not be found for index 3; aborting reading through tracks [?=69 ms]
d:/test.m4b | Disk stream operation : Shortening (delta=-1) [?=9168 ms]
```

Mediainfo is back to

```
General
Complete name                            : D:\test.m4b
Format                                   : MPEG-4
Format profile                           : Base Media
Codec ID                                 : isom (isom/iso2/mp41)
File size                                : 491 MiB
```

**Comment by Zeugma440** [comment]
Thank you so much for the detailed output. I'm gonna look at that later next week.

I'm sorry this all requires so much fine-tuning and discussions. Anyway, I guarantee we'll succeed no matter what ! 💪 

***

`I am not sure why but the first time I save it takes a good amount of time. Further saves even after restarting the program seem faster`

That's Windows' and/or .NET's disk cache at work 😉 

***
`I assume the trak atom errors are from chapters?`

Don't worry, these ones are not errors but information level logs (should even be debug now that I think about it). I should update the output so that the log level appears.

**Comment by ElmoFuntz** [comment]
No problem I don't mind at all. Thanks!

**Comment by Zeugma440** [comment]
- I've detected and fixed one special case where MP4/M4A headers would not be updated properly
- I have fixed the bug where the library would forget the values you set before calling `theTrack.EmbeddedPictures` (yes, that was a silly one)

Here is the updated version

[ATL-20200127.zip](https://github.com/Zeugma440/atldotnet/files/4119505/ATL-20200127.zip)

If it still doesn't work properly on your side, please send me the file you're testing with.

I've just run the library against 4 different M4B's from different sources and it works fine.

**Comment by ElmoFuntz** [comment]
Thanks. I will give it a shot but might be tomorrow. I've got a late day today.

**Comment by ElmoFuntz** [comment]
So good news/bad news. The good news I tested saving, adding a chapter and saving, and updating artwork and the 3 standard test fields and everything seems to be working in VLC. I did crash VLC a few times but I think that is because it doesn't handle 2 chapters having the same start time very well. 

The bad news. iTunes doesn't look at the Nero chapters if its got MP4 chapters (and i'm not positive it will if it doesn't). I tried stripping the chapters out using mp3tag and just writing nero chapters via my program but it did not write anything.

Also FYI I did notice one line in your example code does not work and i'm not sure what this is expecting

`ch.Url = "ccc\0ddd0";`

Gives cannot implicitly convert type 'string' to 'ATL.ChapterInfo.UrlInfo'

Code:

```
         private void btnSave_Click(object sender, EventArgs e)
        {
            // Adapt this to whatever your file path needs to be
            string fileName = "d:/test.m4b";

            IProgress<float> progress = new Progress<float>(displayProgress);
            // Load audio file information into memory
            Track theTrack = new Track(fileName, progress);

            // Modify metadata
            theTrack.Artist = "Hey ho";
            theTrack.Composer = "Oscar Wilde";
            theTrack.Album = "Fake album starts here and is longer than the original one";

            if (theTrack.EmbeddedPictures.Count > 0 && picAlbumArt.Image != null)
            {
                theTrack.EmbeddedPictures.Clear();
                ImageConverter converter = new ImageConverter();
                byte[] data = (byte[])converter.ConvertTo(picAlbumArt.Image, typeof(byte[]));
                PictureInfo newPicture = PictureInfo.fromBinaryData(data, PictureInfo.PIC_TYPE.Front);
                theTrack.EmbeddedPictures.Add(newPicture);
            }

            //Settings.ID3v2_tagSubVersion = 3;
            //Settings.ID3v2_forceUnsynchronization = true;

            ChapterInfo ch = new ChapterInfo();
            ch.StartTime = 6202519;
            ch.StartOffset = 6202519;
            ch.EndTime = 7002519;
            ch.EndOffset = 7002519;
            ch.UniqueID = "";
            ch.Title = "malarky43";
            ch.Subtitle = "bobs your uncle now";

            // Save modifications on the disc

            theTrack.Save();
        }
```

**Comment by Zeugma440** [comment]
So we're doing progress.

I just realized that part of the sample code about Urls is outdated. I'll update it later today. Instanciating a new UrlInfo is the way to go.

***

Regarding iTunes, are you positive nothing happened on the file after saving Nero chapters on an empty file ? Did you check if a `chpl` atom has been added? That part should definitely work.

The scenario I fear the most is the one where the Nero chapters have really been added and iTunes ignore them entirely, waiting for bona fide QT chapters. Creating and editing these will require much work. I can't say it's impossible because I hate saying that, but it's definitely raising the bar in terms of complexity.

Please keep me updated on that.

**Comment by ElmoFuntz** [comment]
Ok so part of the above is an oops on my part. When testing stripping all the chapters and only adding nero chapters bit I also accidentally removed the   theTrack.Chapters.Add(ch); line so yeah nothing was written.  Now for even worse news. With just nero chapters VLC also does not recognize chapters.

I am positive on the other results though and you are correct nero chapters are added and iTunes ignores them. See attached.

![VLC](https://user-images.githubusercontent.com/13905627/73269579-eff38a00-4191-11ea-8e5b-0568481d0795.png)
![itunes](https://user-images.githubusercontent.com/13905627/73269586-f4b83e00-4191-11ea-9cb5-ed4a4030dcf8.png)




Here is Mediainfo's output

Before adding a chapter (I removed a bunch of chapters to make it less of a wall of text)

```
Menu #1
ID                                       : 2
Codec ID                                 : text
Duration                                 : 18 h 1 min
Language                                 : English
Menu For                                 : 1
00:00:00.000                             : Chapter 1
00:25:46.402                             : Chapter 2
00:43:07.631                             : Chapter 3
.. snip ..
17:14:54.224                             : Chapter 47
17:38:03.241                             : Chapter 48
17:55:07.473                             : Chapter 49
Bit rate mode                            : VBR

Menu #2
00:00:00.000                             : Chapter 1
00:25:46.402                             : Chapter 2
00:43:07.631                             : Chapter 3
.. snip ...
17:14:54.224                             : Chapter 47
17:38:03.241                             : Chapter 48
17:55:07.473                             : Chapter 49

```


after adding a chapter

```
Menu #1
ID                                       : 2
Codec ID                                 : text
Duration                                 : 18 h 1 min
Language                                 : English
Menu For                                 : 1
00:00:00.000                             : Chapter 1
00:25:46.402                             : Chapter 2
00:43:07.631                             : Chapter 3
.. snip ..
17:14:54.224                             : Chapter 47
17:38:03.241                             : Chapter 48
17:55:07.473                             : Chapter 49
Bit rate mode                            : VBR



Menu #2
00:00:00.000                             : Chapter 1
00:25:46.402                             : Chapter 2
00:43:07.631                             : Chapter 3
.. snip ..
17:14:54.224                             : Chapter 47
17:38:03.241                             : Chapter 48
17:55:07.473                             : Chapter 49
15:36:42.519                             : malarky43
```


 
after stripping all chapters and adding only nero

```
Text
ID                                       : 2
Format                                   : Timed Text
Codec ID                                 : text
Duration                                 : 18 h 1 min
Bit rate mode                            : Variable
Bit rate                                 : 0 b/s
Stream size                              : 628 Bytes (0%)
Language                                 : English
Default                                  : No
Forced                                   : No

Menu
15:36:42.519                             : malarky43
```

Not sure if this matters but when listing chapters added with my program the output is different see below for an example. The last line with malarky on it is the chapter added notice the time is on the same line. All the rest are below.

Before
```
Chapter 47
(62094224)
Chapter 48
(63483241)
Chapter 49
(64507473)
```

After
```
Chapter 47
(62094224)
Chapter 48
(63483241)
Chapter 49
(64507473)
malarky43(56202519)
```

**Comment by Zeugma440** [comment]
I've also confirmed on my side that MediaPlayer, VLC and foobar2000 all handle both Nero and QT chapters. Only iTunes is ignoring Nero chapters entirely.

***

I don't think `menu#1` vs. `menu#2` is important. That must be a display issue.

***

Have you tried putting timecodes in the chronological order ? Here 15:36:42.519 comes after 17:55:07.473 in the list order, whereas it should come before according to the timecodes. Maybe that's why VLC is crashing ?

***

On my side, the only remaining issue is the fact that ATL should support editing QT chapters in addition to Nero chapters.

I'm opening a new issue for that (see #53).

If you're satisfied with Nero chapters support, you may close the present issue.

**Comment by ElmoFuntz** [comment]
I don't think that's why because VLC does show it correctly in the chapter list but yes I am going to test that. Right now I was just doing quick and dirty tests. To put it in the correct order is more involved and will require I read in all the chapters and sort/insert in the correct location. I will also have to clear and rewrite all of the chapters. All doable just not worked on it yet. I'm not super concerned that VLC crashes since everything else doesn't. I brought it up just incase it might indicate a structure issue. (Getting cannot read memory location errors). I have not tried this on a mobile device yet. I will do that tonight and close this issue after if no major issues.

**Comment by ElmoFuntz** [comment]
Ok just to be sure there was not some remnant left behind from stripping the old chapters out of my test file I created a new m4b and tried adding just nero chapters and it does appear that something gets left in the old file that makes VLC not recognize the nero chapters because my test file does see the chapter i added. Unfortunately iTunes still does not. 

Also i'm sure I did something wrong but your sample code for adding chapters does not work as is and when I removed the offending bits (duplicate ChapterInfo ch = new ChapterInfo();)  it wrote all 3 chapters with the final chapters information. 

```
Audio
ID                                       : 1
Format                                   : ALAC
Codec ID                                 : alac
Codec ID/Info                            : Apple Lossless Audio Codec
Duration                                 : 1 h 6 min
Duration_LastFrame                       : -22 ms
Bit rate mode                            : Variable
Bit rate                                 : 707 kb/s
Nominal bit rate                         : 1 536 kb/s
Channel(s)                               : 2 channels
Sampling rate                            : 48.0 kHz
Bit depth                                : 16 bits
Stream size                              : 334 MiB
Default                                  : Yes
Alternate group                          : 1

Menu
00:01:42.519                             : malarky3 / malarky3 / malarky3
```
![foobar](https://user-images.githubusercontent.com/13905627/73318976-dc770c00-41ef-11ea-8a7a-dd014bd7e85c.png)

Code again

```
         private void btnSave_Click(object sender, EventArgs e)
        {
            // Adapt this to whatever your file path needs to be
            string fileName = "d:/test.m4b";

            IProgress<float> progress = new Progress<float>(displayProgress);
            // Load audio file information into memory
            Track theTrack = new Track(fileName, progress);

            // Modify metadata
            theTrack.Artist = "Hey ho";
            theTrack.Composer = "Oscar Wilde";
            theTrack.Album = "Fake album starts here and is longer than the original one";

            if (theTrack.EmbeddedPictures.Count > 0 && picAlbumArt.Image != null)
            {
                theTrack.EmbeddedPictures.Clear();
                ImageConverter converter = new ImageConverter();
                byte[] data = (byte[])converter.ConvertTo(picAlbumArt.Image, typeof(byte[]));
                PictureInfo newPicture = PictureInfo.fromBinaryData(data, PictureInfo.PIC_TYPE.Front);
                theTrack.EmbeddedPictures.Add(newPicture);
            }

            //Settings.ID3v2_tagSubVersion = 3;
            //Settings.ID3v2_forceUnsynchronization = true;
            theTrack.Chapters = new System.Collections.Generic.List<ChapterInfo>();

            ChapterInfo ch = new ChapterInfo();
            ch.StartTime = 0;
            ch.StartOffset = 0;
            ch.EndTime = 12519;
            ch.EndOffset = 12519;
            ch.UniqueID = "";
            ch.Title = "malarky1";
            ch.Subtitle = "bobs your uncle now";
            theTrack.Chapters.Add(ch);

            //ChapterInfo ch = new ChapterInfo();
            ch.StartTime = 12519;
            ch.StartOffset = 12519;
            ch.EndTime = 102519;
            ch.EndOffset = 102519;
            ch.UniqueID = "";
            ch.Title = "malarky2";
            ch.Subtitle = "bobs your uncle now";
            theTrack.Chapters.Add(ch);


            //ChapterInfo ch = new ChapterInfo();
            ch.StartTime = 102519;
            ch.StartOffset = 102519;
            ch.EndTime = 3002519;
            ch.EndOffset = 3002519;
            ch.UniqueID = "";
            ch.Title = "malarky3";
            ch.Subtitle = "bobs your uncle now";
            theTrack.Chapters.Add(ch);


            // Save modifications on the disc

            theTrack.Save();
        }
```

**Comment by Zeugma440** [comment]
Sample code in the wiki looks fine. Just look closely : we don't repeat `ChapterInfo ch = new ChapterInfo()`, we instanciate a new chapter with `ch = new ChapterInfo()`. That should fix your "all chapters have final chapter info" bug. 

***

Regarding Nero chapters and VLC, let me rephrase the remaining issue you found
1. Pick an m4b without any chapters
2. Add Nero chapters with ATL
3. Open with VLC

Expected : Chapters are recognized by VLC
Actual : VLC does not recognize chapters 

Is that it?

**Comment by ElmoFuntz** [comment]
Ok my copy/paste skills and proofreading failed me :) That works.

Regarding the other part its more a nonissue and more information for you if you need it since previously I had said VLC was not seeing chapters also.

- Create an m4b without chapters
- Add nero chapters with ATL

Result VLC and foobar2000 see chapters. iTunes does not


- Pick an M4B that has both chapter types and strip them with mp3tag
- Add nero chapters with ATL

Result only foobar2000 sees chapters

**Comment by Zeugma440** [comment]
```
- Create an m4b without chapters
- Add nero chapters with ATL
Result : VLC and foobar2000 see chapters. iTunes does not
```
That issue will be fixed in #53 as iTunes exclusively reads QT chapters.
No matter what we do with Nero chapters, that won't fix anything with iTunes.

***

```
- Pick an M4B that has both chapter types and strip them with mp3tag
- Add nero chapters with ATL
Result only foobar2000 sees chapters
```

I just did exactly these steps with 
- the M4B file you sent me (_Trial and Flame : Chronicle book 2_) 
- the code you posted this morning

Result : VLC, Media Player and foobar2000 manage to display all three "malarky" chapters. VLC doesn't crash either when quitting.

However, as the file is quite large (356 MB), VLC takes 3 to 4 seconds before displaying the chapters on the progress bar. Maybe that's why you though they weren't there ?

Last but not least, the problem you reported might be specific to the file you're testing with. If you definitely confirm VLC can't see the chapters you added, please send me the original file you're testing with. Thanks !

Versions used : MP3tag v2.94, VLC v3.0.6, fb2k v1.4.3, Media Player Classic Home Cinema v1.7.8 162

**Comment by ElmoFuntz** [comment]
Ok not sure what happened but its working now with VLC as well. Can you remove that title from your last post? Rather not have it archived :) Can close after.

**Comment by Zeugma440** [comment]
Done. Thanks for your patience and for all the constructive feedback.

I'm working on QT chapters at the moment. No roadblock so far.
